publish/src/main/java/com/indix/gocd/s3publish/PublishExecutor.java

```diff
@@ -142,17 +146,45 @@ private ExecutionResult envNotFound(String environmentVariable) {
         return ExecutionResult.failure(message);
     }
 
-    private void setMetadata(GoEnvironment env, String bucket, S3ArtifactStore store) {
+    private void setMetadata(GoEnvironment env, String bucket, String destinationPrefix, S3ArtifactStore store) {
         ObjectMetadata metadata = metadata(env);
         metadata.setContentLength(0);
         InputStream emptyContent = new ByteArrayInputStream(new byte[0]);
         PutObjectRequest putObjectRequest = new PutObjectRequest(bucket,
-                env.artifactsLocationTemplate() + "/",
+                ensureKeySegmentValid(destinationPrefix),
                 emptyContent,
                 metadata);
 
         store.put(putObjectRequest);
     }
+
+    private String getDestinationPrefix(final TaskConfig config, final GoEnvironment env) {
+        String destinationPrefix = config.getValue(DESTINATION_PREFIX);
+
+        if(org.apache.commons.lang3.StringUtils.isBlank(destinationPrefix)) {
+            return env.artifactsLocationTemplate();
+        }
+
+        destinationPrefix = env.replaceVariables(destinationPrefix);
+
+        if(destinationPrefix.endsWith("/")) {
+            destinationPrefix = destinationPrefix.substring(0, destinationPrefix.length() - 1);
+        }
+
+        return destinationPrefix;
+    }
+
+    private String ensureKeySegmentValid(String segment) {
+        if(org.apache.commons.lang3.StringUtils.isBlank(segment)) {
+            return segment;
+        }
+
+        if(!org.apache.commons.lang3.StringUtils.endsWith(segment, "/")) {
```

Comment:
I think I just used this as I saw it done this way elsewhere in the project. Most of my experience is with C#, I've only worked on a few small Java projects. As a result, I'm not too sure on the best practice in regards to static import vs import. I've just imported StringUtils (not statically) as that is how ArrayUtils is imported.